<a href="https://colab.research.google.com/github/VasilisVar00/artificial_intelligence_ntua/blob/main/programming/AI_21_EX2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Ομάδα 

Στοιχεία Μέλους 1: Βασίλειος Βαρσαμής - 03118033

Στοιχεία Μέλους 2: Αλέξανδρος Κουλάκος - 03118144

# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής 
14. Ίδια δεκαετία 

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata):
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5). 

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θα έπρεπε να έχει προταθεί στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να επιστρέφεται 1 ή 0 αν η ταινία θα έπαιρνε βαθμό μεγαλύτερο του 3 ή όχι, δηλαδή αν θα έπρεπε να την είχαμε προτείνει ή όχι.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). 

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις. 

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί. 

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/content/drive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/content/drive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά. 

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import os
# path="/content/drive/MyDrive/AI_EX2_2021/" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας
path="/content/drive/MyDrive/" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας

Mounted at /content/drive


In [ ]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [ ]:
import pandas as pd
from pyswip import Prolog

In [ ]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd

# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv(path + "movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()


Unnamed: 0     budget                                    genres  \
0           0  237000000  Action|Adventure|Fantasy|Science Fiction   
1           1  300000000                  Adventure|Fantasy|Action   
2           2  245000000                    Action|Adventure|Crime   
3           3  250000000               Action|Crime|Drama|Thriller   
4           4  260000000          Action|Adventure|Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                       plot_keywords  language  \
0  culture clash|future|space war|space colony|so...   English   
1  ocean|drug abuse|exotic island|east india trad...   English   
2  spy|based on novel|secret agent|sequel|mi6|bri...  Français   
3  dc comics|crime fighter|terrorist|secret ident...   English   
4  based on novel|mars|medallion|space travel|pri...   English   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ...  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   
3  Following the death of District Attorney Harve...  112.312950  ...   
4  John Carter is a war-weary, former military ca...   43.926995  ...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                movie_title vote_average  num_voted_users  \
0                                    Avatar          7.2            11800   
1  Pirates of the Caribbean: At World's End          6.9             4500   
2                                   Spectre          6.3             4466   
3                     The Dark Knight Rises          7.6             9106   
4                               John Carter          6.1             2124   

  title_year                   country      director_name     actor_1_name  \
0     2009.0  United States of America      James Cameron      Zoe Saldana   
1     2007.0  United States of America     Gore Verbinski    Orlando Bloom   
2     2015.0            United Kingdom         Sam Mendes  Christoph Waltz   
3     2012.0  United States of America  Christopher Nolan    Michael Caine   
4     2012.0  United States of America     Andrew Stanton     Lynn Collins   

       actor_2_name       actor_3_name  
0  Sigourney Weaver       Stephen Lang  
1   Keira Knightley  Stellan Skarsgård  
2       Léa Seydoux      Ralph Fiennes  
3       Gary Oldman      Anne Hathaway  
4   Samantha Morton       Willem Dafoe  

[5 rows x 27 columns]

In [ ]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [ ]:
# Returns a list with the names of production companies 
# String s has the format of a 
# list of dictionaries.
def convert_company(s):
    myList = []
    
    s = s[1 : -1]
    s = s.replace("{", "")
    s = s.replace("}", "")
    s = s.replace("name: ", "")
    s = s.replace("id: ", "")
    
    # print(s)
    
    helpy = s.split(", ")
    
    length = len(helpy)
    
    i = 0
    
    while i < length:
        if not(helpy[i].isnumeric()) and (helpy[i + 1].isnumeric()):
            myList.append(helpy[i])
            i += 2
            continue
        elif not(helpy[i].isnumeric()) and not(helpy[i + 1].isnumeric()):
            myList.append(helpy[i] + ", " + helpy[i + 1])
            i += 3
            continue
        i += 1
    
    return myList

# Returns a list with the names of production countries 
# String s has the format of a 
# list of dictionaries.
def convert_country(s):
    myList = []
    
    s = s[1 : -1]
    s = s.replace("{", "")
    s = s.replace("}", "")
    s = s.replace("iso_3166_1: ", "")
    s = s.replace("name: ", "")
    
    # print(s)
    
    helpy = s.split(", ")
    
    length = len(helpy)
    
    for i in range(1, length, 2):
        myList.append(helpy[i])
    
    return myList

In [ ]:
# create World
# Ορίζουμε τον κόσμο μας
prolog = Prolog()
# prolog.query("make")

# Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκεύσουμε
# αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}

for row in data.itertuples(index=True, name='Pandas'):
  colour = "Yes"

  movie_title = clean_text(getattr(row, 'movie_title'))
  genres = clean_text(getattr(row, 'genres'))
  keywords = clean_text(getattr(row, 'plot_keywords'))
  director_name = clean_text(getattr(row, 'director_name'))
  title_year = clean_text(str(getattr(row, 'title_year')))
  duration = clean_text(str(getattr(row, 'duration')))
  language = clean_text(getattr(row, 'language'))
  production_companies_string = clean_text(getattr(row, 'production_companies'))
  production_countries_string = clean_text(getattr(row, 'production_countries'))

  if movie_title != "UNK":

    literals.append("movie('"+ movie_title +"')")

    if genres != "UNK":
      for genre in genres.split("|"):
        literals.append("genre('"+ movie_title +"','"+ genre +"')")

    if keywords != "UNK":
      for keywords in keywords.split("|"):
        literals.append("keywords('"+ movie_title +"','"+ keywords +"')")

        if keywords == "black and white":
          colour = "NO"
    
      literals.append("colour('"+ movie_title +"','"+ colour +"')")
    
    if director_name != "UNK":
      literals.append("director('"+ movie_title +"', '"+ director_name +"')")

    acc = []

    actor_1 = clean_text(getattr(row, 'actor_1_name'))
    if actor_1 != "UNK":
      acc.append(actor_1)

    actor_2 = clean_text(getattr(row, 'actor_2_name'))
    if actor_2 != "UNK":
      acc.append(actor_2)

    actor_3 = clean_text(getattr(row, 'actor_3_name'))
    if actor_3 != "UNK":
      acc.append(actor_3)

    for actors in acc:
      literals.append("actors('"+ movie_title +"', '"+ actors.replace("'", '"') +"')")

    if title_year != "UNK":
      literals.append("year('"+ movie_title +"', '"+ title_year +"')")

    if duration != "UNK":
      literals.append("duration('"+ movie_title +"', '"+ duration +"')")

    if language != "UNK":
      literals.append("language('"+ movie_title +"', '"+ language +"')")

    if production_companies_string != "[]":
      production_companies_list = convert_company(production_companies_string)
      for item in production_companies_list:
        literals.append("studio('"+ movie_title +"', '"+ item +"')")
    
    if production_countries_string != "[]":
      production_countries_list = convert_country(production_countries_string)
      for item in production_countries_list:
        literals.append("country('"+ movie_title +"', '"+ item +"')")

# Η Prolog θέλει τα κατηγορήματά της με την σειρά 
literals.sort()
for literal in literals:
  prolog.assertz(literal)
#   print (literal +'.')


#Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
#prolog.consult(path + "db.pl")

In [ ]:
# Returns the Jaccard similarity of lists L1 and L2 which contain the genres of movies M1 and M2
prolog.assertz('(jaccard_similarity_genres(M1, M2, D) :- movie(M1), movie(M2), \+ (M1 = M2), create_genres_list(M1, L1), create_genres_list(M2, L2), union(L1, L2, U), intersection(L1, L2, I), length(U, X), length(I, Y), D is Y / X)')

# Creates a list of all genres that are related to movie M
prolog.assertz('(create_genres_list(M, L) :- findall(G, genre(M, G), L))')

# ------ Query 1 ------- #

# We are making the hypothesis that two movies have the same topic 
# iff the Jaccard similarity of their genres is greater than or equal to 0.6.
prolog.assertz('(similar(M1, M2):- jaccard_similarity_genres(M1, M2, D), D >= 0.6)')

# ------ Query 2 ------- #

# We are making the hypothesis that two movies have the same topic 
# iff the Jaccard similarity of their genres is greater than or equal to 0.3 and less than 0.6.
prolog.assertz('(pretty_similar(M1, M2):- jaccard_similarity_genres(M1, M2, D), D < 0.6, D >= 0.3)')

# ------ Query 3 ------- #

# We are making the hypothesis that two movies have the same topic 
# iff the Jaccard similarity of their genres is greater than or equal to 0.1 and less than 0.3.
prolog.assertz('(relatively_similar(M1, M2):- jaccard_similarity_genres(M1, M2, D), D < 0.3, D >= 0.1)')

# Creates a list of all keywords that are related to movie M
prolog.assertz('(create_keywords_list(M, L) :- findall(K, keywords(M, K), L))')

# ------ Query 4 ------- #

prolog.assertz('(same_director(M1, M2) :- director(M1, D), director(M2, D), \+ (M1 = M2))')

# ------ Query 5 ------- #

prolog.assertz('(same_plot(M1, M2) :-movie(M1),movie(M2),create_keywords_list(M1, L1),create_keywords_list(M2, L2),intersection(L1, L2, I),length(I, L),L >= 3)')

# ------ Query 6 ------- #

prolog.assertz('(similar_plot(M1, M2) :- movie(M1), movie(M2), create_keywords_list(M1, L1), create_keywords_list(M2, L2), intersection(L1, L2, I),length(I, L),L < 3, L > 0)')

# ------ Query 7 ------- #

prolog.assertz('(three_same_actors(X, Y) :- actors(X, A), actors(X, B), actors(X, C), actors(Y, A), actors(Y, B), actors(Y, C), \+ (X = Y), \+ (A = B), \+ (A = C), \+ (B = C))')

# ------ Query 8 ------- #

prolog.assertz('(two_same_actors(X, Y) :- actors(X, A), actors(X, B), actors(Y, A), actors(Y, B), \+ (X = Y), \+ (A = B), \+ (three_same_actors(X, Y)))')

# ------ Query 9 ------- #

prolog.assertz('(one_same_actor(X, Y) :- actors(X, A), actors(Y, A), \+ (X = Y), \+ (two_same_actors(X, Y)))')

# ------ Query 10 ------- #

prolog.assertz('(same_language(M1, M2) :- language(M1, L), language(M2, L), \+ (M1 = M2))')

# ------ Query 11 ------- #

prolog.assertz('(same_colour(M1, M2) :- colour(M1, C), colour(M2, C), \+ (M1 = M2))')

# ------ Query 12 ------- #

prolog.assertz('(same_studio(M1, M2) :- studio(M1, S), studio(M2, S), \+ (M1 = M2))')

# ------ Query 13 ------- #

prolog.assertz('(same_country(M1, M2) :- country(M1, C), country(M2, C), \+ (M1 = M2))')

# ------ Query 14 ------- #

# Given a year, return the decade it belongs to
prolog.assertz('(decade(M, D) :- year(M, Aux), atom_number(Aux, NewY), Y is floor(NewY),  D is Y mod 100 - Y mod 10)')

# Given a year, return the century it belongs to
prolog.assertz('(century(M, C) :- year(M, Aux), atom_number(Aux, NewY), Y is floor(NewY), C is Y mod 1000 - Y mod 100)')

# Two movies belong to the same decade if first and foremost belong to the same century
# For example, movie with title year 1916 must not belong to the same decade as the movie with title year 2016.
prolog.assertz('(same_decade(M1, M2) :- century(M1, C), century(M2, C), decade(M1, D), decade(M2, D), \+ (M1 = M2))')



In [ ]:
# ------------------------------------------------------------------------ #
#                                 Testing
# ------------------------------------------------------------------------ #

def print_sol(query, key):
  s = set()

  for x in query:
    s.add(x['M'])
  
  print(key, list(s))

def test(movie):
  query = prolog.query("similar('" + movie + "', M)")
  print_sol(query, "similar")
  query.close

  query = prolog.query("pretty_similar('" + movie + "', M)")
  print_sol(query, "pretty_similar")
  query.close

  query = prolog.query("relatively_similar('" + movie + "', M)")
  print_sol(query, "relatively_similar")
  query.close

  query = prolog.query("same_plot('" + movie + "', M)")
  print_sol(query, "same_plot")
  query.close

  query = prolog.query("similar_plot('" + movie + "', M)")
  print_sol(query, "similar_plot")
  query.close

  query = prolog.query("similar('" + movie + "', M)")
  print_sol(query, "similar")
  query.close

  query = prolog.query("three_same_actors('" + movie + "', M)")
  print_sol(query, "three_same_actors")
  query.close

  query = prolog.query("two_same_actors('" + movie + "', M)")
  print_sol(query, "two_same_actors")
  query.close

  query = prolog.query("one_same_actor('" + movie + "', M)")
  print_sol(query, "one_same_actor")
  query.close

  query = prolog.query("same_director('" + movie + "', M)")
  print_sol(query, "same_director")
  query.close

  query = prolog.query("same_language('" + movie + "', M)")
  print_sol(query, "same_language")
  query.close

  query = prolog.query("same_colour('" + movie + "', M)")
  print_sol(query, "same_colour")
  query.close

  query = prolog.query("same_studio('" + movie + "', M)")
  print_sol(query, "same_studio")
  query.close

  query = prolog.query("same_country('" + movie + "', M)")
  print_sol(query, "same_country")
  query.close

  query = prolog.query("same_decade('" + movie + "', M)")
  print_sol(query, "same_decade")
  query.close

In [ ]:
movie = "The Avengers"

test(movie)

similar ['The Empire Strikes Back', 'Raiders of the Lost Ark', 'Star Trek: First Contact', 'Babylon A.D.', 'Indiana Jones and the Last Crusade', 'Star Trek: Nemesis', 'Waterworld', 'Star Wars: Episode I - The Phantom Menace', 'Lucy', 'Independence Daysaster', 'Underworld: Rise of the Lycans', 'The Matrix Reloaded', 'Armageddon', 'Spy Kids 3-D: Game Over', 'AVP: Alien vs. Predator', 'Iron Man', 'Sheena', 'The Wolverine', 'Firefox', 'Star Trek II: The Wrath of Khan', 'Mystery Men', 'G.I. Joe: Retaliation', 'Shin Godzilla', 'Superman Returns', 'Krrish', 'Universal Soldier: The Return', 'After Earth', 'Dragonball Evolution', 'Star Wars', 'Superman', 'Ant-Man', 'Steel', 'The Lost World: Jurassic Park', 'Superman IV: The Quest for Peace', 'Divergent', 'The Phantom', 'Stargate: The Ark of Truth', 'Starship Troopers', 'The Fifth Element', 'Star Trek: Generations', 'Star Trek: Insurrection', 'X-Men Origins: Wolverine', 'Men in Black II', 'Terminator Genisys', 'Beastmaster 2: Through the Portal 

# Σχολιασμός Μέρους 1

*   Στο αρχείο movies_metadata.csv που μας δόθηκε, τα attributes 
production_companies και production_countries έχουν το format λιστών που έχουν ως στοιχεία dictionaries εκ των οποίων μας ενδιαφέρει μόνο το value που αντιστοιχεί στο name ως key. Ωστόσο, η συνάρτηση clean_text() επιστρέφει την πληροφορία των production_companies και production_countries ως string. Έτσι, αξιοποιώντας το "βολικό" format αυτών των string, χρησιμοποιούμε τις συναρτήσεις convert_country και convert_company για να απομονώσουμε τα ονόματα των χωρών και εταιρειών παραγωγής τα οποία αποθηκεύουμε τελικά στη βάση της Prolog.

*   Για να καθορίσουμε την ομοιότητα ανάμεσα στα genres ταινιών χρησιμοποιούμε το συντελεστή ομοιότητας $Jaccard$ ο οποίος περιγράφει την ποσοστιαία εγγύτητα δύο συνόλων $Α$, $Β$ και ορίζεται ως εξής: $J = \frac{|Α \cap Β|}{|A \cup B|}$. Πράγματι, αυτό είναι ένα αντικειμενικό κριτήριο ομοιότητας, διότι στα genres κάθε ταινίας εμφανίζονται λίγες λέξεις, οπότε η ποσοστιαία σύγκρισή τους δίνει μια καλύτερη εικόνα για τη θεματική τους εγγύτητα. Συγκεκριμένα, για να είναι δύο ταινίες κοντινές ως προς τη θεματολογία τους, δεν αρκεί να έχουν 2 - 3 κοινές λέξεις στο genre, διότι π.χ. η ταινία $M_1$ με $genre =  romance$ $|$ $comedy$ είναι θεματικά παραπλήσια με την ταινία $Μ_2$ με $genre = romance$. Αυτό, όμως, δε φαίνεται από το πλήθος των κοινών λέξεων το οποίο εν προκειμένω ισούται με τη μονάδα, φαίνεται καλύτερα από το συντελεστή $Jaccard$ ο οποίος ισούται με $\frac{2}{3}\ = 0.66$). Αντίθετα, η ομοιότητα ταινιών ως προς τα plot keywords εκτιμάται αποδοτικότερα, αν μετρήσουμε το πλήθος των κοινών keywords. Αναλυτικά, λόγω των πολλών λέξεων που εμφανίζονται στα keywords, 3 - 4 κοινές λέξεις αρκούν για να αποφανθούμε ότι πράγματι οι δύο ταινίες έχουν παρόμοια πλοκή.

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


In [ ]:
prolog.assertz('(find_sim_5(M1, M2) :- similar(M1, M2), same_plot(M1, M2), same_language(M1, M2))')
prolog.assertz('(find_sim_5(M1, M2) :- similar(M1, M2), similar_plot(M1, M2), same_director(M1, M2), same_language(M1, M2))')
prolog.assertz('(find_sim_5(M1, M2) :- similar(M1, M2), similar_plot(M1, M2), one_same_actor(M1, M2), same_language(M1, M2))')
prolog.assertz('(find_sim_5(M1, M2) :- similar(M1, M2), two_same_actors(M1, M2), same_language(M1, M2))')
prolog.assertz('(find_sim_5(M1, M2) :- three_same_actors(M1, M2), same_language(M1, M2))')

prolog.assertz('(find_sim_4(M1, M2) :- similar(M1, M2), similar_plot(M1, M2))')
prolog.assertz('(find_sim_4(M1, M2) :- similar(M1, M2), one_same_actor(M1, M2))')
prolog.assertz('(find_sim_4(M1, M2) :- similar(M1, M2), same_director(M1, M2))')

prolog.assertz('(find_sim_3(M1, M2) :- pretty_similar(M1, M2), similar_plot(M1, M2))')
prolog.assertz('(find_sim_3(M1, M2) :- pretty_similar(M1, M2), one_same_actor(M1, M2), same_studio(M1, M2))')
prolog.assertz('(find_sim_3(M1, M2) :- pretty_similar(M1, M2), two_same_actors(M1, M2))')
prolog.assertz('(find_sim_3(M1, M2) :- pretty_similar(M1, M2), same_director(M1, M2))')

prolog.assertz('(find_sim_2(M1, M2) :- pretty_similar(M1, M2), same_colour(M1, M2), same_decade(M1, M2))')
prolog.assertz('(find_sim_2(M1, M2) :- pretty_similar(M1, M2), same_studio(M1, M2))')
prolog.assertz('(find_sim_2(M1, M2) :- one_same_actor(M1, M2))')
prolog.assertz('(find_sim_2(M1, M2) :- same_director(M1, M2))')

prolog.assertz('(find_sim_1(M1, M2) :- relatively_similar(M1, M2), same_decade(M1, M2), same_colour(M1, M2))')


In [ ]:
# Function to find the movies returned by the query (the function returns a set)
def find_sol(query): 
  s = set()
  for sol in query:
    m = sol['M']
    if m not in s:
      s.add(m)
  return s

def simple_recommender(movie):
    q = prolog.query("find_sim_5('" + movie +"',M)")
    s = find_sol(q)
    answer_5 = set()
    if(s):
      answer_5 = s
      
    q = prolog.query("find_sim_4('" + movie +"',M)")
    s = find_sol(q)
    answer_4 = set()
    if(s):
      answer_4 = s
    
    q = prolog.query("find_sim_3('" + movie +"',M)")
    s = find_sol(q)
    answer_3 = set()
    if(s):
      answer_3 = s

    q = prolog.query("find_sim_2('" + movie +"',M)")
    s = find_sol(q)
    answer_2 = set()
    if(s):
      answer_2 = s
              
    q = prolog.query("find_sim_1('" + movie +"',M)")
    s = find_sol(q)
    answer_1 = set()
    if(s):
      answer_1 = s
             
    q.close()

    result = []
    seen = set()
    count = 6;
    for i in [answer_5, answer_4, answer_3, answer_2, answer_1]:
      count -= 1
      if(i):
        for j in i:
          if j not in seen:
            result.append((count,j))
            seen.add(j)
    return result

In [ ]:
movie = "Gladiator"

simple_recommender(movie)[:25]

[(4, 'Apocalypto'),
 (4, 'Kingdom of Heaven'),
 (4, 'Pompeii'),
 (4, 'Warrior'),
 (4, 'Exodus: Gods and Kings'),
 (4, 'Centurion'),
 (4, 'Never Back Down'),
 (4, 'White Squall'),
 (4, 'Robin Hood'),
 (4, 'Flight of the Intruder'),
 (4, '1911'),
 (4, 'G.I. Jane'),
 (3, '1982'),
 (3, 'Planet of the Apes'),
 (3, 'Alexander'),
 (3, 'Super'),
 (3, 'Return of the Jedi'),
 (3, 'The Martian'),
 (3, 'Star Trek: Nemesis'),
 (3, 'The Birth of a Nation'),
 (3, 'Star Wars: Episode I - The Phantom Menace'),
 (3, 'The Promise'),
 (3, 'One Mans Hero'),
 (3, 'Star Wars: Episode II - Attack of the Clones'),
 (3, 'Crossroads')]

# Σχολιασμός Μέρους 2

Όπως φαίνεται και στο παράδειγμα που εκτελείται στο παραπάνω κελί, οι προτεινόμενες ταινίες εκτυπώνονται σε φθίνουσα σειρά ως προς το βαθμό ομοιότητας, έτσι ώστε για κάθε ταινία που προτείνεται να αναγράφεται και ο βαθμός ομοιότητας με την ταινία που δόθηκε ως input (ο βαθμός ομοιότητας χρειάζεται στο Μέρος 3). Στις κλάσεις ομοιότητας 4 - 5 ανήκουν οι ταινίες με κοινή θεματολογία, παραπλήσια πλοκή και κοινούς ηθοποιούς ή σκηνοθέτη. Οι κανόνες αυτοί εξασφαλίζουν ότι προτείνονται από τον recommender ταινίες πολύ κοντινές σε αυτή που δόθηκε ως input (π.χ. για $movie = Star$ $Trek$ εμφανίζονται όλα τα sequels, καθώς και όλες οι ταινίες Star Wars). Στις κλάσεις ομοιότητας 1 - 3 συμπεριλάβαμε λιγότερους και πιο "χαλαρούς" κανόνες που σχετίζονται με τη δεκαετία, το χρώμα και το studio παραγωγής. Τέλος, σημειώνουμε ότι δεν έχουμε τρόπο να ελέγξουμε με ποια σειρά εμφανίζονται ταινίες που ανήκουν στην ίδια κλάση ομοιότητας, διότι αυτές επιλέγονται από το διερμηνέα της Prolog σύμφωνα με τη σειρά που εμφανίζονται στη βάση της. 

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [ ]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i for i in range(1,6)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie) # our recommender returns [(similarity,similar_movie)]

        for similarity,similar_movie in similar_movies:
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[similarity]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[similarity] #το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


#αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict
#έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

Η παραπάνω εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος στάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε κάθε φορά να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτόν να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας αν είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου και αν είχαμε επιλέξει διαφορετικές 3 ταινίες να τα πηγαίναμε χάλια. Οπότε προτείνουμε να τρέξετε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών π.χ. 10 πειράματα με 3 ταινίες, 10 πειράματα για 20 ταινίες κ.ο.κ. και να κρατήσετε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα των αγγλικών σας θα είναι πιο αντικειμενικά.

In [ ]:
train_ratings = pd.read_csv(path + "AI_EX2_2021/train_ratings.csv")
test_ratings = pd.read_csv(path + "AI_EX2_2021/test_ratings.csv")

In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5141692194561234
recall: 0.6652777777777776
f1: 0.5792917953436757


In [ ]:
metrics = []
for i in range(10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 20)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5165025066864785
recall: 0.7138888888888889
f1: 0.5934918363178381


In [ ]:
metrics = []
for i in range(10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5148770553430264
recall: 0.7875
f1: 0.6217651593825532


Εξετάζουμε τις μετρικές που παίρνουμε, όταν εκπαιδεύσουμε τον recommender με το σύνολο όλων των ταινιών που περιέχονται στο αρχείο train_ratings.csv(100 ταινίες). Αυτό μας δίνει ένα άνω φράγμα για τις μετρικές που πετυχαίνουμε αν εκπαιδέυσουμε τον recommender με ένα υποσύνολο ταινιών του αρχείου.

In [ ]:
#train with the whole dataset, this is the upper bound  score of our recommender
metrics = []

movie_score = train_recommender(train_ratings, rating_weights, score_weights, -1)
real, pred = predict_example(test_ratings, movie_score)
metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

precision: 0.5149253731343284
recall: 0.9583333333333334
f1: 0.6699029126213593


# Σχολιασμός Μέρους 3

Παρατηρούμε ότι το precision είναι γύρω στο 50%, αυτό σημαίνει ότι πάνω από τις μισές ταινίες προτείνονται σωστά, ταυτόχρονα σημαίνει ότι λίγο λιγότερες από τις μισές ταινίες που προτείνονται, προτείνονται λάθος. Το precision με λίγα λόγια μας δίνει μια εικόνα για το πόσες ταινίες είναι falsely predicted. Όπως βλέπουμε και παραπάνω, το precision δεν διαφέρει σημαντικά όταν εκπαιδεύουμε με κάποιο υποσύνολο ταινιών σε σχέση με όλες τις ταινίες, αφού σε κάθε περίπτωση είναι ανάμεσα στο 50% και το 52%. Από την άλλη, το recall μας δίνει μια εικόνα για το πόσες ταινίες είναι falsely not predicted, δηλαδή πόσες ταινίες δεν προτείναμε ενώ θα έπρεπε. Είναι λογικό το recall να αυξάνεται όσο παίρνουμε μεγαλύτερα υποσύνολα, γιατί έτσι παράγουμε καινούριες προτεινόμενες ταινίες με βάση κάποιο rating ταινίας που δεν ήταν στο σύνολο μικρότερου μεγέθους. Παρατηρούμε ότι για 10 ταινίες το recall είναι γύρω στο 50% που σημαίνει ότι 1 στις 2 ταινίες που θα έπρεπε να προταθεί δεν προτείνεται. Το ποσοστό αυξάνει όσο μεγαλώνουμε το υποσύνολο με αποτέλεσμα στις 30 ταινίες να έχουμε γύρω στο 75% recall, δηλαδή 1 στις 4 ταινίες δεν προτείνεται ενώ θα έπρεπε. Το recall παίρνει την μέγιστη τιμή του όταν εκπαιδεύουμε με όλο το training_set, από το οποίο παίρνουμε το πολυ ικανοποιητικό ποσοστό του 95.8%.